[RISCV] Add searchable table for tune information

wangpc-pp opened the pull request:
There are many information that can be used for tuning, like
alignments, cache line size, etc. But we can't make all of them
`SubtargetFeature` because some of them are not with enumerable
value, for example, `PrefetchDistance` used by `LoopDataPrefetch`.

In this patch, a searchable table `RISCVTuneInfoTable` is added,
in which each entry contains the CPU name and all tune information
defined in `RISCVTuneInfo`. Each field of `RISCVTuneInfo` should
have a default value and processor definitions can override the
default value via `let` statements.

We don't need to define a `RISCVTuneInfo` for each processor and
it will use the default value (which is for `generic`) if no
`RISCVTuneInfo` defined.

For processors in the same series, a subclass can inherit from
`RISCVTuneInfo` and override the fields. And we can also override
the fields in processor definitions if there are some differences
in the same processor series.

When initilizing `RISCVSubtarget`, we will use `TuneCPU` as the
key to serach the tune info table. So, the behavior here is if
we don't specify the tune CPU, we will use specified `CPU`, which
is expected I think. 

This patch almost undoes 61ab106, in which I added tune features
of preferred function/loop alignments. More tune information can
be added in the future.

wangpc-pp committed:
[RISCV] Add searchable table for tune information

There are many information that can be used for tuning, like
alignments, cache line size, etc. But we can't make all of them
`SubtargetFeature` because some of them are not with enumerable
value, for example, `PrefetchDistance` used by `LoopDataPrefetch`.

In this patch, a searchable table `RISCVTuneInfoTable` is added,
in which each entry contains the CPU name and all tune information
defined in `RISCVTuneInfo`. Each field of `RISCVTuneInfo` should
have a default value and processor definitions can override the
default value via `let` statements.

When initilizing `RISCVSubtarget`, we will use `TuneCPU` as the
key to serach the tune info table. So, the behavior here is if
we don't specify the tune CPU, we will use specified `CPU`, which
is expected I think.

This patch almost undoes 61ab106, in which I added tune features
of preferred function/loop alignments. More tune information can
be added in the future.

michaelmaitland commented on llvm/lib/Target/RISCV/RISCVProcessors.td:
```diff
@@ -10,20 +10,36 @@
                           list<SubtargetFeature> tunef = [],
                           string default_march = "">
-      :  ProcessorModel<n, m, f, tunef> {
+      :  ProcessorModel<n, m, f, tunef>, RISCVTuneInfo {
```
Should `RISCVTuneInfo` be an optional argument to `ProcessorModel`? Should `RISCVTuneInfo` be an optional argument to `RISCVProcessorModel`?

If we organize it like this, it becomes more clear to clients of `ProcessorModel` and `RISCVProcessorModel` that they have the ability to specify these tunings.

topperc replied:
What would overriding these for a specific CPU look like right now. Would we need to add `let` for each field to each CPU?

I'd kind of like a way to define common sets that can be used by multiple CPUs without being repeating.

wangpc-pp replied:
I changed the implementation.
We don't need to define a `RISCVTuneInfo` for each processor and it will use the fault value (which is for `generic`) if no `RISCVTuneInfo` defined.
For processors in the same series, a subclass can inherit from `RISCVTuneInfo` and override the fields. And we can also override the fields in processor definitions if there are some differences in the same processor series.
It can be like:
```
class SiFive7TuneInfo: RISCVTuneInfo {
  let XXX = xxx; // Override the field in class definition.
  // ...
};

def SIFIVE_7 : RISCVTuneProcessorModel<"sifive-7-series",
                                       SiFive7Model,
                                       [TuneSiFive7]>, SiFive7TuneInfo;

def SIFIVE_U74 : RISCVProcessorModel<"sifive-u74",
                                     SiFive7Model,
                                     [Feature64Bit,
                                      FeatureStdExtZifencei,
                                      FeatureStdExtM,
                                      FeatureStdExtA,
                                      FeatureStdExtF,
                                      FeatureStdExtD,
                                      FeatureStdExtC],
                                     [TuneSiFive7]>, SiFive7TuneInfo {
  let XXX = xxx; // Override the field in processor definition.
  // ...
}
```

michaelmaitland commented on llvm/test/CodeGen/RISCV/align.ll:
```diff
@@ -2,8 +2,6 @@
 ; RUN:   | FileCheck %s -check-prefix=RV32I
 ; RUN: llc -mtriple=riscv32 -mattr=+c -verify-machineinstrs < %s \
 ; RUN:   | FileCheck %s -check-prefix=RV32C
-; RUN: llc -mtriple=riscv32 -mattr=+pref-func-align-32 -verify-machineinstrs < %s \
```
Are there any processors that have non-default alignment? We could pass -mcpu and check for the `.p2align N` line so that we have test coverage.

wangpc-pp replied:
I don't know one. Maybe you can provide more information about SiFive processors in upstream?

michaelmaitland replied:
We haven't noticed any performance issues from instruction alignment on SiFive7 and are relying on linker to get this right. I'm not so sure what else we can do to have test coverage for this without an upstream processor that uses this TuneInfo 🤷

michaelmaitland replied:
I've just learned that different alignment values have different results on different workloads for us. Therefore, I think we should have way to control this value at compile time.

topperc replied:
There are IR attributes that override the CPU defaults. falign-loops and falign-functions uses those attributes.

wangpc-pp replied:
I found some descriptions in https://starfivetech.com/uploads/u74mc_core_complex_manual_21G1.pdf:
> 3.2.6 Instruction Fetch Unit
> The S7 instruction fetch unit is responsible for keeping the pipeline fed with instructions from
memory. The instruction fetch unit delivers up to 8 bytes of instructions per clock cycle to sup-
port superscalar instruction execution. _Fetches are always **word-aligned** and there is a one-
cycle penalty for branching to a 32-bit instruction that is not word-aligned._

It seems we'd better to align functions and hot loops to 4-byte to get best performance. But `C` extension is supported so the alignment would be 2-byte.
```
  // Function alignments.
  const Align FunctionAlignment(Subtarget.hasStdExtCOrZca() ? 2 : 4);
  setMinFunctionAlignment(FunctionAlignment);
  // Set preferred alignments.
  setPrefFunctionAlignment(Subtarget.getPrefFunctionAlignment());
  setPrefLoopAlignment(Subtarget.getPrefLoopAlignment());
```

topperc commented on llvm/lib/Target/RISCV/RISCVProcessors.td:
```diff
@@ -10,20 +10,36 @@
 // RISC-V processors supported.
 //===----------------------------------------------------------------------===//
 
+class RISCVTuneInfo {
```
Should these fields be in RISCVProcessorModel if we're always going to have RISCVProcessorModel inherit this class?

wangpc-pp committed:
fixup! [RISCV] Add searchable table for tune information

Define RISCVTuneInfo by need and fail back to generic if no related RISCVTuneInfo existed

wangpc-pp commented:
Ping.

topperc commented:
> We collected some data that shows that different values of these variables are better on different benchmarks. Do we have the ability to override these values at compile time so that different benchmarks can specify which alignment to use?

Alignment can be overridden by frontend options `-falign-loops` and `-falign-functions`. This sets an attribute on each function or loop in IR. if those attributes are present, they are used instead of the information in TargetLowering/Subtarget.

wangpc-pp commented:
> > We collected some data that shows that different values of these variables are better on different benchmarks. Do we have the ability to override these values at compile time so that different benchmarks can specify which alignment to use?
> 
> Alignment can be overridden by frontend options `-falign-loops` and `-falign-functions`. This sets an attribute on each function or loop in IR. if those attributes are present, they are maxed with the information in TargetLowering/Subtarget.

Besides, user can set alignment per function/loop via `__attribute__(aligned(n))` or something like `__attribute__((optimize("align-loops=n")))`.

topperc commented on llvm/lib/Target/RISCV/RISCVSubtarget.cpp:
```diff
@@ -65,6 +71,11 @@ RISCVSubtarget::initializeSubtargetDependencies(const Triple &TT, StringRef CPU,
+  TuneInfo = RISCVTuneInfoTable::getRISCVTuneInfo(TuneCPU);
+  // If there is no TuneInfo for this CPU, we fail back to generic.
+  if (!TuneInfo)
+    TuneInfo = RISCVTuneInfoTable::getRISCVTuneInfo("generic");
```
assert than `TuneInfo` isn't null for "generic"?

topperc reviewed (approved):
LGTM other than that comment.

wangpc-pp committed:
fixup! [RISCV] Add searchable table for tune information

Add assertion
